exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/aggregate/HashAggBatch.java

```diff
@@ -149,14 +149,24 @@ public IterOutcome innerNext() {
       if ( aggOut == HashAggregator.AggIterOutcome.AGG_OK ) { return IterOutcome.OK; }
       // if RESTART - continue below with doWork() - read some spilled partition, just like reading incoming
       incoming = aggregator.getNewIncoming(); // Restart - incoming was just changed
-      if ( wasKilled ) { // if kill() was called before, then finish up
-        aggregator.cleanup();
-        incoming.kill(false);
-        return IterOutcome.NONE;
-      }
     }
 
-    AggOutcome out = aggregator.doWork();
+    if ( wasKilled ) { // if kill() was called before, then finish up
+      aggregator.cleanup();
+      incoming.kill(false);
+      return IterOutcome.NONE;
+    }
+
+    // Read and aggregate records
+    // ( may need to run again if the spilled partition that was read
+    //   generated new partitions that were all spilled )
+    AggOutcome out = AggOutcome.CALL_WORK_AGAIN;
+    while ( out == AggOutcome.CALL_WORK_AGAIN) {
+      //
+      //  Read incoming batches and process their records
+      //
+      out = aggregator.doWork();
+    }
```

Comment:
```
while (aggregator.doWork() == AggOutcome.CALL_WORK_AGAIN) {
  // Nothing to do
}
```
?

In one of your reviews you said you didn't like empty loops, but sometimes they are handy...